server/live_games/game.py

```diff
@@ -1,191 +1,143 @@
 
 class GameLogic:
-    def __init__(self, game_id):
-        self.game_id = game_id
-        # Init game maze
-        self._screen_size = {'height': screen_height, 'width': screen_width}
-        self._paddle_size = paddle_size
-        # Init paddles
-        self._player1: Paddle = Paddle('Left', self._screen_size.get('height'), self._paddle_size) # left player
-        self._player2: Paddle = Paddle('Right', self._screen_size.get('height'), self._paddle_size) # right player
-        # Init ball
-        self._ball: Ball = Ball({'N': True, 'S': True, 'W': False, 'E': False}, self._screen_size) # Set walls to True
-
-        self._start_time = time.time()
-        self._last_tick = 0
+    def __init__(self, game_id, player1_username, player2_username):
+
+        self._game_id = game_id
+        self._player1_username = player1_username
+        self._player2_username = player2_username
+        self._map_width = 1000
+        self._map_height = 500 
+        self._paddle_size = 4
+        self._ball_speed = 1
+        self._initial_countdown_value = 3
+        self._current_countdown = self._initial_countdown_value
         self._player1_ready = False
         self._player2_ready = False
+        self._phase = "Waiting players"
+        self._start_time = -1
+        self._last_tick = 0
+        self._player1: Paddle = Paddle('Left', self._map_height, self._paddle_size)
+        self._player2: Paddle = Paddle('Right', self._map_height, self._paddle_size)
+        self._ball: Ball = Ball(self._map_width, self._map_height, self._ball_speed)
 
     def move_player1(self, direction: int):
         self._player1.move_paddle(direction)
+
     def move_player2(self, direction: int):
         self._player2.move_paddle(direction)
 
     def set_player1_ready(self):
         self._player1_ready = True
         if self._player2_ready:
-            self.start()
+            asyncio.create_task(self.render_game())
+
     def set_player2_ready(self):
         self._player2_ready = True
         if self._player1_ready:
-            self.start()
-
-    def start(self):
-        self._start_time = time.time() + 3
+            asyncio.create_task(self.render_game())
+
+    async def render_game(self):
+        self._start_time = time.time() + self._initial_countdown_value
+        self._phase = "countdown"
+        await self.render_countdown()
+        self._phase = "running"
+        await self.render_ball_movement()
+
+    async def render_countdown(self):
+        while self._current_countdown > 0:
+            new_countdown = self._start_time - time.time()
+            if new_countdown > 0:
+                self._current_countdown = new_countdown
+            else:
+                self._current_countdown = 0
+            await asyncio.sleep(0.016) # 60 fps
+
+    async def render_ball_movement(self):
+        while self._phase == "running":
+            self._ball.movement(self._player1, self._player2)
+            await asyncio.sleep(0.016) # 60 fps
```

Comment:
Having these sleeps everywhere doesn't look v safe tbh, also it would be good to have the 'engine' running in a separate thread. I'd suggest putting all the functions you want to run on an interval into 1 wrapper function and use threading.Timer(). Also you don't need to worry about being 60fps. The canvas rerenders at whatever the refresh rate of the monitor is, and shows whatever the state represents on each cycle. You can just make the interval whatever seems to be the most performant. Will approve change though, and you can decide if it's fine or if you want to change later

Reply:
I let chat gpt o take care of this .. hahah 🤣 

1. "Having these sleeps everywhere doesn't look very safe"

This part reflects a concern that the use of sleep() functions (presumably asyncio.sleep()) might introduce problems, such as blocking or race conditions. However, in the context of asynchronous programming, using **asyncio.sleep()** is safe and appropriate.

    Why asyncio.sleep() is safe: It is a non-blocking way to pause the execution of a specific coroutine while allowing the rest of the event loop to continue running other tasks. It does not block the main event loop, and this is the intended way to handle time-based delays in an asynchronous framework like asyncio.

    Where the confusion might come from: The reviewer might be conflating blocking time.sleep() with the non-blocking asyncio.sleep(). It's common for developers unfamiliar with asyncio to worry about the implications of "sleep" calls, but in this context, asyncio.sleep() is perfectly safe and appropriate.

2. "It would be good to have the 'engine' running in a separate thread"

Running the "engine" (presumably game logic or rendering logic) in a separate thread isn't necessary when you're already using asyncio. In fact, mixing threads with asyncio when you don't have CPU-bound tasks can introduce unnecessary complexity and even performance bottlenecks.

    Why asyncio is better than threads here:
        asyncio provides concurrency without needing threads. By running asynchronous coroutines concurrently, you avoid the overhead and complexity of managing threads and synchronization mechanisms like locks or semaphores.
        Introducing threads can lead to race conditions, especially if multiple threads are accessing shared state (e.g., game state). asyncio's event loop avoids this by handling concurrency cooperatively.

3. "I'd suggest putting all the functions you want to run on an interval into one wrapper function and use threading.Timer()"

This suggestion is problematic because **threading.Timer()** is designed for blocking, synchronous code. It does not integrate well with asyncio, which is non-blocking and designed for concurrent I/O-bound tasks.

    Why threading.Timer() isn’t a good idea:
        threading.Timer() introduces threads for what is essentially an event-driven task. This can add unnecessary complexity to the code and lead to race conditions if shared resources (like the game state) are not properly managed with thread synchronization.
        asyncio.sleep() already handles periodic intervals in an elegant and non-blocking way. Wrapping functions with threading.Timer() is overkill and introduces more complexity without solving a real problem.

4. "You don't need to worry about being 60fps"

This part of the comment is accurate in some ways but misses the purpose of setting an interval for updates (like 60 FPS in your case):

    Why you might want to control the update frequency:
        Even though the canvas rendering in a browser can match the monitor's refresh rate (e.g., using requestAnimationFrame()), the game state still needs to be updated on a fixed interval, especially for real-time games. This is independent of the canvas rendering.
        Updating the game state at roughly 60 FPS (or a reasonable rate) ensures that the ball moves, collisions are detected, and user inputs are handled in a timely manner. Without a controlled update interval, the game logic might be inconsistent or feel sluggish.
    Why 60 FPS is a reasonable choice:
        While the rendering might depend on the monitor refresh rate, the game logic should still run at a consistent frame rate (like 60 FPS) to ensure smooth, predictable physics and gameplay.
        Canvas rendering vs. game logic: You can decouple the canvas rendering rate from the game update rate, but for most games, a 60 FPS game update interval works well with standard refresh rates (60Hz, 120Hz, etc.).

Suggestions for a More Informed Comment:

    Keep using asyncio.sleep() instead of time.sleep() or threading.Timer(). It’s non-blocking and fits perfectly into the asynchronous event loop of asyncio.

    Don’t introduce threads unless necessary. Threads would complicate the design unnecessarily, and asyncio is designed to handle concurrency for I/O-bound tasks in a simpler, more predictable way.

    Control the update rate (e.g., 60 FPS) independently of the canvas rendering. Even if the rendering adjusts dynamically, the game logic should run at a consistent rate for smooth updates.

Revised PR Comment Example:

If I were leaving feedback based on what you’ve shared, here’s how I might rephrase the PR comment to make it more aligned with best practices:

    Feedback: The use of asyncio.sleep() is appropriate here, as it allows the event loop to continue processing other tasks while managing game logic updates at the desired frame rate. Introducing threads via threading.Timer() would add complexity without clear benefits, and asyncio already provides concurrency in a more controlled manner.

    Regarding the 60 FPS update interval, while the canvas renders according to the monitor's refresh rate, maintaining a consistent update rate for the game logic (e.g., 60 FPS) ensures smooth gameplay and reliable game state updates. You don’t need to worry about overloading the CPU, as asyncio.sleep() ensures this remains non-blocking and lightweight